File: service/app/mastery.py
"""Mastery model — single source of truth, mirrored EXACTLY by
web/src/lib/mastery.ts. Both are tested against shared/mastery-fixtures.json;
change one, change all three.

Pure functions only: no DB, no clock — callers supply elapsed days.
"""

from __future__ import annotations

# rating (1-4, FSRS Again/Hard/Good/Easy) → concept-score delta
REVIEW_DELTAS = {1: -0.15, 2: 0.02, 3: 0.08, 4: 0.12}

DECAY_HALF_LIFE_DAYS = 90.0


def clamp01(x: float) -> float:
    return 0.0 if x < 0.0 else 1.0 if x > 1.0 else x


def apply_review(score: float, rating: int) -> float:
    """Concept score update from one parsing-card review (or checker result
    mapped onto the same 1-4 scale)."""
    return clamp01(score + REVIEW_DELTAS.get(rating, 0.0))


def apply_decay(score: float, days: float) -> float:
    """Gentle time decay: half-life 90 days."""
    if days <= 0:
        return score
    return clamp01(score * 0.5 ** (days / DECAY_HALF_LIFE_DAYS))


def lesson_mastered(
    deck_stats: list[dict],
    concept_scores: list[float],
    retention_threshold: float = 0.9,
    concept_threshold: float = 0.8,
) -> bool:
    """A lesson is mastered when every one of its decks has all cards at FSRS
    review state with retention >= threshold, AND every one of its concepts
    scores >= threshold. A lesson with no decks can't be mastered by data
    (use the manual override — Jess outranks the model of her own mastery)."""
    if not deck_stats:
        return False
    for deck in deck_stats:
        if not deck.get("all_review"):
            return False
        if (deck.get("retention") or 0.0) < retention_threshold:
            return False
    for score in concept_scores:
        if score < concept_threshold:
            return False
    return True

File: web/src/lib/mastery.ts
/* Framework-free (Divergence Rule 2): no React, no fetch, no DOM.
   Unit-testable in plain Node.

   Mastery model — EXACT mirror of service/app/mastery.py (the source of
   truth). Both are tested against shared/mastery-fixtures.json; change one,
   change all three. */

export const REVIEW_DELTAS: Record<number, number> = { 1: -0.15, 2: 0.02, 3: 0.08, 4: 0.12 }

export const DECAY_HALF_LIFE_DAYS = 90

export function clamp01(x: number): number {
  return x < 0 ? 0 : x > 1 ? 1 : x
}

/** Concept score update from one parsing-card review. */
export function applyReview(score: number, rating: number): number {
  return clamp01(score + (REVIEW_DELTAS[rating] ?? 0))
}

/** Gentle time decay: half-life 90 days. */
export function applyDecay(score: number, days: number): number {
  if (days <= 0) return score
  return clamp01(score * 0.5 ** (days / DECAY_HALF_LIFE_DAYS))
}

export type DeckMasteryStats = { all_review: boolean; retention: number | null }

/** Mastered = every deck fully at review state with retention >= threshold,
 *  AND every concept >= threshold. No decks → never mastered by data. */
export function lessonMastered(
  deckStats: DeckMasteryStats[],
  conceptScores: number[],
  retentionThreshold = 0.9,
  conceptThreshold = 0.8,
): boolean {
  if (deckStats.length === 0) return false
  for (const deck of deckStats) {
    if (!deck.all_review) return false
    if ((deck.retention ?? 0) < retentionThreshold) return false
  }
  for (const score of conceptScores) {
    if (score < conceptThreshold) return false
  }
  return true
}
